staging/cse/windows/azurecnifunc.ps1

```diff
@@ -494,6 +494,79 @@ function GenerateAzureStackCNIConfig
     Set-ItemProperty -Path $azureCNIConfigFile -Name IsReadOnly -Value $true
 }
 
+
+function GetIpv4AddressFromParsedContent
+{
+    param (
+        [Parameter(Mandatory = $true)]
+        $ParsedContent
+    )
+
+    if ($ParsedContent[0].ipv4 -and $ParsedContent[0].ipv4.ipAddress -and $ParsedContent[0].ipv4.ipAddress.Count -gt 0)
+    {
+        return $ParsedContent[0].ipv4.ipAddress[0].privateIpAddress
+    }
+    else
+    {
+        return $null
+    }
+}
+
+function GetIpv6AddressFromParsedContent
+{
+    param (
+        [Parameter(Mandatory = $true)]
+        $ParsedContent
+    )
+
+    if ($ParsedContent[0].ipv6 -and $ParsedContent[0].ipv6.ipAddress -and $ParsedContent[0].ipv6.ipAddress.Count -gt 0)
+    {
+        return $ParsedContent[0].ipv6.ipAddress[0].privateIpAddress
+    }
+    else
+    {
+        return $null
+    }
+}
+
+function GetMetadataContent
+{
+    $Retries = 10
+    $RetryDelaySeconds = 15
+
+    for ($i = 0; $i -lt $Retries; $i++) {
+        try
+        {
+            $MetadataContent = Invoke-WebRequest -UseBasicParsing -Uri "http://169.254.169.254/metadata/instance/network/interface?api-version=2021-02-01" -Headers @{ "metadata" = "true" } -TimeoutSec 10 -ErrorAction Stop
+            $ParsedContent = $MetadataContent.Content | ConvertFrom-Json
+            $ipv4Address = GetIpv4AddressFromParsedContent -ParsedContent $ParsedContent
+            if (-not $ipv4Address)
+            {
+                Write-Log "Failed to retrieve IPv4 address from metadata. Will retry in $RetryDelaySeconds seconds. Attempt $( $i + 1 ) of $Retries."

```

Comment:
Is there a reason to retry on parse errors? Do you expect malformed response?

Reply:
IMDS sometimes doesn't come up in time so we need to retry. I'm not sure if IMDS gives a 200 response with bad data in these case or a 5xx response, so retry in both cases makes sense.